File: ios/SiliconBuddy/Transport/TransportError.swift
import Foundation

/// Why a call to the Mac did not answer.
///
/// The distinctions here are the ones a person can act on. "Unreachable" means turn
/// Tailscale on; "connection refused" means check the listener and port; "unauthorized"
/// means pair again. Collapsing those into one "network error" would make the app
/// useless exactly when something is wrong.
public enum TransportError: Error, Equatable, Sendable {
    /// Nothing at that address answered: wrong host, tailnet down, Mac asleep.
    case unreachable(String)
    /// The connection was refused. The app may be closed, or the address/port may be wrong.
    case appNotRunning
    /// The request took too long to get anywhere.
    case timedOut
    /// 401. The token is wrong, or the Mac revoked this device.
    case unauthorized
    /// 403. The token is good but this device may not do that — a chat-scope device
    /// asking to load a model, or a wrong pairing code. The Mac's own words matter
    /// here, because "not allowed" and "no longer paired" call for different actions.
    case forbidden(String)
    /// 409. The Mac is already answering in that conversation.
    case conflict(String)
    /// 413. More than a device may send: 4 MiB a body, about 1.5 MB an image.
    case tooLarge(String)
    /// 411. The Mac will not read a chunked body. This one is ours to fix, not the
    /// owner's: every request this app sends sets a Content-Length.
    case chunkedNotAccepted(String)
    /// 404 on a route that exists, about a thing that does not — a conversation that
    /// was deleted on the Mac, say. Not the same as "this Mac is too old".
    case notFound(String)
    /// 404 on a route this app knows about — which for the M0 routes means "that Mac
    /// has not shipped them yet", and the caller should fall back rather than fail.
    case routeUnavailable(String)
    /// 400 with the Mac's own explanation; the most common one is "no model is loaded".
    case badRequest(String)
    /// 415: not a picture or a clip the Mac will keep. Only `POST /uploads` answers it.
    case unsupportedMedia(String)
    /// 416: a `Range` outside the file. Only `GET /media/{id}` answers it.
    case rangeNotSatisfiable(String)
    /// 429: the Mac is already doing as much of this as it will do at once.
    case busy(String)
    /// 503: the Mac cannot reach something it needs for this right now — the drive its
    /// model library lives on, for the models it keeps for the phone. Its sentence names it.
    case unavailable(String)
    /// 507: the Mac has no room for what was asked, said before anything moved.
    case insufficientStorage(String)
    /// Any other HTTP status, with whatever the Mac said.
    case server(status: Int, message: String)
    /// The body did not match the contract.
    case decoding(String)
    /// No Mac has been paired yet.
    case notConfigured
    /// The caller went away.
    case cancelled
}

extension TransportError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .unreachable(let host):
            "Can't reach \(host). Check that Tailscale is on and the Mac is awake."
        case .appNotRunning:
            "Nothing is listening at this address and port. Open Silicon Optimizer and check the address and port."
        case .timedOut:
            "The Mac took too long to answer."
        case .unauthorized:
            "This device isn't paired any more. Pair it again from the Mac."
        case .forbidden(let message):
            message.isEmpty ? "The Mac wouldn't allow that." : message
        case .conflict(let message):
            message.isEmpty
                ? "That conversation is still being answered." : message
        case .tooLarge(let message):
            message.isEmpty ? "That was too large to send." : message
        case .chunkedNotAccepted(let message):
            message.isEmpty ? "The Mac wouldn't read that request." : message
        case .notFound(let message):
            message.isEmpty ? "The Mac doesn't have that any more." : message
        case .routeUnavailable(let path):
            "This Mac doesn't have \(path) yet."
        case .badRequest(let message):
            message
        case .unsupportedMedia(let message):
            message.isEmpty ? "That file is not one this Mac will keep." : message
        case .rangeNotSatisfiable(let message):
            message.isEmpty ? "That part of the file is not there any more." : message
        case .busy(let message):
            message
        case .unavailable(let message):
            message.isEmpty ? "The Mac can't do that right now." : message
        case .insufficientStorage(let message):
            message.isEmpty ? "The Mac has no room for that." : message
        case .server(let status, let message):
            message.isEmpty ? "The Mac returned an error (\(status))." : message
        case .decoding(let detail):
            "The Mac's answer didn't match what this app expects: \(detail)"
        case .notConfigured:
            "No Mac is paired yet."
        case .cancelled:
            "Cancelled."
        }
    }

    /// What to offer the person, when there is something to offer.
    public var recoverySuggestion: String? {
        switch self {
        case .unreachable: "Open Tailscale, then pull to refresh."
        case .appNotRunning: "Open Silicon Optimizer and check the address and port."
        case .unauthorized: "Settings → Silicon Buddy → Pair a device."
        case .forbidden: "Pair again from the Mac with full control."
        case .conflict: "Wait for the answer, or start another conversation."
        case .tooLarge: "Send fewer or smaller pictures."
        case .notFound: "It may have been deleted on the Mac."
        case .badRequest: "Load a model from the Models tab."
        case .insufficientStorage: "Free some space on the Mac, then try again."
        default: nil
        }
    }

    /// True when a caller should quietly use its fallback instead of showing this.
    public var isMissingRoute: Bool {
        if case .routeUnavailable = self { return true }
        return false
    }

    /// A 404 about a thing rather than a route: the same status, a different meaning,
    /// and only the caller knows which it asked for.
    public var asNotFound: TransportError {
        if case .routeUnavailable(let path) = self {
            return .notFound("The Mac has nothing at \(path) any more.")
        }
        return self
    }

    /// True when the Mac refused because of what this device is allowed to do, rather
    /// than because it does not know this device.
    public var isForbidden: Bool {
        if case .forbidden = self { return true }
        return false
    }
}

extension TransportError {
    /// Maps what URLSession and the server say into the vocabulary above.
    ///
    /// `NSURLErrorCannotConnectToHost` is the interesting one: the TCP connection was
    /// refused. That alone cannot distinguish a closed app from the wrong address or port.
    public static func from(urlError error: URLError) -> TransportError {
        switch error.code {
        case .cancelled:
            return .cancelled
        case .timedOut:
            return .timedOut
        case .cannotConnectToHost:
            return .appNotRunning
        case .appTransportSecurityRequiresSecureConnection:
            // Only reachable if the app's ATS exception is ever narrowed: it means the
            // system refused plain HTTP to that address, not that the Mac is away.
            return .forbidden(
                "iOS refused a plain HTTP connection to that address. "
                    + "Use the Mac's Tailscale name or address."
            )
        case .cannotFindHost, .dnsLookupFailed, .notConnectedToInternet,
             .networkConnectionLost, .internationalRoamingOff, .dataNotAllowed,
             .secureConnectionFailed, .resourceUnavailable:
            let host = error.failingURL?.host ?? "the Mac"
            return .unreachable(host)
        default:
            let host = error.failingURL?.host ?? "the Mac"
            return .unreachable(host)
        }
    }

    /// Maps an HTTP status plus the Mac's error body.
    public static func from(status: Int, body: Data, path: String) -> TransportError? {
        guard !(200..<300).contains(status) else { return nil }
        let message = (try? JSONDecoder().decode(ControlAPI.ErrorResponse.self, from: body))?.error
            ?? String(data: body, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? ""
        switch status {
        case 401: return .unauthorized
        case 403: return .forbidden(message)
        case 404: return .routeUnavailable(path)
        case 409: return .conflict(message)
        case 411: return .chunkedNotAccepted(message)
        case 413: return .tooLarge(message)
        case 400: return .badRequest(message.isEmpty ? "The Mac rejected the request." : message)
        case 415: return .unsupportedMedia(message)
        case 416: return .rangeNotSatisfiable(message)
        case 429: return .busy(message.isEmpty ? "The Mac is busy." : message)
        case 503: return .unavailable(message)
        case 507: return .insufficientStorage(message)
        default: return .server(status: status, message: message)
        }
    }
}

File: android/app/src/main/java/dev/siliconoptimizer/buddy/transport/TransportError.kt
package dev.siliconoptimizer.buddy.transport

import kotlinx.serialization.json.Json
import java.io.IOException
import java.net.ConnectException
import java.net.NoRouteToHostException
import java.net.SocketTimeoutException
import java.net.UnknownHostException

/**
 * Why a call to the Mac did not answer.
 *
 * The distinctions here are the ones a person can act on. "Unreachable" means turn
 * Tailscale on; "the app isn't running" means open Silicon Optimizer; "unauthorized"
 * means pair again. Collapsing those into one "network error" would make the app
 * useless exactly when something is wrong.
 */
sealed class TransportError(message: String) : Exception(message) {
    /** Nothing at that address answered: wrong host, tailnet down, Mac asleep. */
    data class Unreachable(val host: String) :
        TransportError("Can't reach $host. Check that Tailscale is on and the Mac is awake.")

    /** The address answered but nothing is listening on the port. */
    data object AppNotRunning :
        TransportError("The Mac answered, but Silicon Optimizer isn't running on it.")

    data object TimedOut : TransportError("The Mac took too long to answer.")

    /** 401: the token is wrong, or the Mac revoked this device. */
    data object Unauthorized :
        TransportError("This device isn't paired any more. Pair it again from the Mac.")

    /**
     * 403: the token is good but this device may not do that — a chat-scope device
     * asking to load a model, or a wrong pairing code. The Mac's own words matter here,
     * because "not allowed" and "no longer paired" call for different actions.
     */
    data class Forbidden(val detail: String) : TransportError(detail)

    /** 409: the Mac is already answering in that conversation. */
    data class Conflict(val detail: String) : TransportError(detail)

    /** 413: more than a device may send — 4 MiB a body, about 1.5 MB an image. */
    data class TooLarge(val detail: String) : TransportError(detail)

    /**
     * 411: the Mac will not read a chunked body. This one is ours to fix, not the
     * owner's: every request this app sends sets a Content-Length.
     */
    data class ChunkedNotAccepted(val detail: String) : TransportError(detail)

    /**
     * 404 on a route that exists, about a thing that does not — a conversation deleted
     * on the Mac, say. Not the same as "this Mac is too old".
     */
    data class NotFound(val detail: String) : TransportError(detail)

    /** 404 on a route this app knows about: that Mac has not shipped it yet. */
    data class RouteUnavailable(val path: String) :
        TransportError("This Mac doesn't have $path yet.")

    /** 400 with the Mac's own explanation; most often "no model is loaded". */
    data class BadRequest(val detail: String) : TransportError(detail)

    /** 429: the Mac is already doing as much of this as it will do at once. */
    data class Busy(val detail: String) : TransportError(detail)

    data class Server(val status: Int, val detail: String) :
        TransportError(if (detail.isBlank()) "The Mac returned an error ($status)." else detail)

    data class Decoding(val detail: String) :
        TransportError("The Mac's answer didn't match what this app expects: $detail")

    data object NotConfigured : TransportError("No Mac is paired yet.")

    data object Cancelled : TransportError("Cancelled.")

    val isMissingRoute: Boolean get() = this is RouteUnavailable

    /**
     * True when the Mac refused because of what this device may do, rather than
     * because it does not know this device.
     */
    val isForbidden: Boolean get() = this is Forbidden

    /**
     * A 404 about a thing rather than a route: the same status, a different meaning,
     * and only the caller knows which it asked for.
     */
    fun asNotFound(): TransportError = when (this) {
        is RouteUnavailable -> NotFound("The Mac has nothing at $path any more.")
        else -> this
    }

    /**
     * What may be written to a log: the shape of the failure and nothing else.
     *
     * `message` is written for the person looking at the screen, and several of these
     * carry the Mac's tailnet address in it — `Unreachable` by construction, `Forbidden`
     * when Android refuses cleartext — while others carry whatever text the Mac put in
     * an error body. logcat is a different audience: it survives the moment, it goes
     * into bug reports, and an address on the owner's tailnet is not a thing to leave
     * lying around in one. So nothing here interpolates a host, a path, a conversation
     * id or any server-supplied text; a status code is as specific as it gets.
     *
     * Kept exhaustive deliberately — a new case has to choose its own tag rather than
     * fall into a branch that might print something.
     */
    val logSummary: String
        get() = when (this) {
            is Unreachable -> "unreachable"
            is AppNotRunning -> "app not running"
            is TimedOut -> "timed out"
            is Unauthorized -> "unauthorized"
            is Forbidden -> "forbidden"
            is Conflict -> "conflict"
            is TooLarge -> "too large"
            is ChunkedNotAccepted -> "chunked not accepted"
            is NotFound -> "not found"
            is RouteUnavailable -> "route unavailable"
            is BadRequest -> "bad request"
            is Busy -> "busy"
            is Server -> "server error $status"
            is Decoding -> "decoding failed"
            is NotConfigured -> "not configured"
            is Cancelled -> "cancelled"
        }

    /** What to offer the person, when there is something to offer. */
    val recovery: String?
        get() = when (this) {
            is Unreachable -> "Open Tailscale, then pull to refresh."
            is AppNotRunning -> "Open Silicon Optimizer on the Mac."
            is Unauthorized -> "Settings, Silicon Buddy, Pair a device."
            is Forbidden -> "Pair again from the Mac with full control."
            is Conflict -> "Wait for the answer, or start another conversation."
            is TooLarge -> "Send fewer or smaller pictures."
            is NotFound -> "It may have been deleted on the Mac."
            is BadRequest -> "Load a model from the Models tab."
            else -> null
        }

    companion object {
        private val lenient = Json { ignoreUnknownKeys = true }

        /**
         * Maps what the socket layer says.
         *
         * [ConnectException] is the interesting one: the connection was actively
         * refused, which means the host is there and the port is not — the Mac is awake
         * and the app is closed.
         */
        fun from(error: IOException, host: String): TransportError = when (error) {
            is SocketTimeoutException -> TimedOut
            is ConnectException -> AppNotRunning
            is UnknownHostException, is NoRouteToHostException -> Unreachable(host)
            else -> {
                val text = error.message.orEmpty()
                when {
                    // The platform refusing plain HTTP is not the Mac being away, and
                    // the fix is different: use the address it was paired with.
                    text.contains("Cleartext", true) -> Forbidden(
                        "Android refused a plain HTTP connection to " + host + ". " +
                            TailnetHost.EXPLANATION,
                    )
                    text.contains("ECONNREFUSED", true) ||
                        text.contains("refused", true) -> AppNotRunning
                    text.contains("timed out", true) -> TimedOut
                    else -> Unreachable(host)
                }
            }
        }

        /** Maps an HTTP status plus the Mac's error body. Null when the call succeeded. */
        fun from(status: Int, body: String, path: String): TransportError? {
            if (status in 200..299) return null
            val detail = runCatching {
                lenient.decodeFromString(ErrorResponse.serializer(), body).error
            }.getOrElse { body.trim() }
            return when (status) {
                401 -> Unauthorized
                403 -> Forbidden(detail.ifBlank { "The Mac wouldn't allow that." })
                404 -> RouteUnavailable(path)
                409 -> Conflict(detail.ifBlank { "That conversation is still being answered." })
                411 -> ChunkedNotAccepted(detail.ifBlank { "The Mac wouldn't read that request." })
                413 -> TooLarge(detail.ifBlank { "That was too large to send." })
                400 -> BadRequest(detail.ifBlank { "The Mac rejected the request." })
                429 -> Busy(detail.ifBlank { "The Mac is busy." })
                else -> Server(status, detail)
            }
        }
    }
}
